src/KOKKOS/metatensor_system_kokkos.cpp

```diff
@@ -0,0 +1,412 @@
+    auto device = system->positions().device();
+
+    auto positions_kokkos = this->atomKK->k_x. template view<DeviceType>();
+    auto total_n_atoms = atomKK->nlocal + atomKK->nghost;
+    
+    /*-------------- this will be done on CPU for now ------------------------*/
+    // There is no kokkos cell in LAMMPS, so we need to transfer
+    auto cell_inv_tensor = system->cell().inverse().t().to(device).to(dtype);
+
+    // The hashmap in the following code is not easy to implement in either Kokkos or torch
+    // The cost of this section seems to be very low anyway
+
+    // Collect the local atom id of all local & ghosts atoms, mapping ghosts
+    // atoms which are periodic images of local atoms back to the local atoms.
+    //
+    // Metatensor expects pairs corresponding to periodic atoms to be between
+    // the main atoms, but using the actual distance vector between the atom and
+    // the ghost.
+    original_atom_id_.clear();
+    original_atom_id_.reserve(total_n_atoms);
+
+    // identify all local atom by their LAMMPS atom tag.
+    local_atoms_tags_.clear();
+    for (int i=0; i<atom->nlocal; i++) {
+        original_atom_id_.emplace_back(i);
+        local_atoms_tags_.emplace(atom->tag[i], i);
+    }
+
+    // now loop over ghosts & map them back to the main cell if needed
+    ghost_atoms_tags_.clear();
+    for (int i=atom->nlocal; i<total_n_atoms; i++) {
+        auto tag = atom->tag[i];
+        auto it = local_atoms_tags_.find(tag);
+        if (it != local_atoms_tags_.end()) {
+            // this is the periodic image of an atom already owned by this domain
+            original_atom_id_.emplace_back(it->second);
+        } else {
+            // this can either be a periodic image of an atom owned by another
+            // domain, or directly an atom from another domain. Since we can not
+            // really distinguish between these, we take the first atom as the
+            // "main" one and remap all atoms with the same tag to the first one
+            auto it = ghost_atoms_tags_.find(tag);
+            if (it != ghost_atoms_tags_.end()) {
+                // we already found this atom elsewhere in the system
+                original_atom_id_.emplace_back(it->second);
+            } else {
+                // this is the first time we are seeing this atom
+                original_atom_id_.emplace_back(i);
+                ghost_atoms_tags_.emplace(tag, i);
+            }
+        }
+    }
+
+    auto original_atom_id_tensor = torch::from_blob(
+        original_atom_id_.data(),
+        {total_n_atoms},
+        torch::TensorOptions().dtype(torch::kInt32).device(torch::kCPU)
+    );
+    original_atom_id_tensor = original_atom_id_tensor.to(device);  // RIP
+
+    /*----------- end of "this will be done on CPU for now" --------------*/
+
+
+    NeighListKokkos<DeviceType>* list_kk = static_cast<NeighListKokkos<DeviceType>*>(this->list_);
+
+    auto numneigh_kk = list_kk->d_numneigh;
```

Comment:
are we sure this is always a `int32_t`, even when lammps is compiled with different integer sizes? https://docs.lammps.org/Build_settings.html#size-of-lammps-integer-types-and-size-limits

Reply:
Yikes, you're right. Since we're converting these ints to torch, either we make up a tupe conversion system or we error out if someone tries to compile these files while LAMMPS's integers are not `int32`. I would go for the second one